lib/Dialect/Arc/Transforms/IsolateClocks.cpp

```diff
@@ -0,0 +1,217 @@
+        clockedOp && clockedOp.isClocked())
+      continue;
+    // Don't pull in operations that have other users outside this domain. We
+    // don't want do duplicate ops to keep binary size as small as possible and
+    // avoid computing the same thing multiple times. This is also important
+    // because we want every path from domain input to domain output to have at
+    // least one cycle latency, ideally every output is directly from a state op
+    // with greater than 0 latency such that we don't have to insert additional
+    // storage slots later on.
+    if (llvm::any_of(op->getUsers(), [&](auto *user) {
+          return user->getBlock() != domainBlock.get();
+        }))
+      continue;
+
+    if (moveToDomain(op))
+      worklist.append(op->getOperands().begin(), op->getOperands().end());
+  }
+}
+
+ClockDomainOp ClockDomain::materialize(OpBuilder &materializeBuilder,
+                                       Location loc) {
+  builder.setInsertionPointToEnd(domainBlock.get());
+  SmallVector<Value> inputs, outputs;
+  computeCrossingValues(inputs, outputs);
+
+  // Add the terminator and clock domain outputs and rewire the SSA value uses
+  auto outputOp = builder.create<arc::OutputOp>(loc, outputs);
+  auto clockDomainOp = materializeBuilder.create<ClockDomainOp>(
+      loc, ValueRange(outputs).getTypes(), inputs, clock);
+  for (auto [domainOutput, val] :
+       llvm::zip(clockDomainOp.getOutputs(), outputOp->getOperands())) {
+    val.replaceUsesWithIf(domainOutput, [&](OpOperand &operand) {
+      return operand.getOwner()->getBlock() != domainBlock.get();
+    });
+  }
+
+  // Add arguments and inputs to the clock domain operation and rewire the SSA
+  // value uses.
+  domainBlock->addArguments(ValueRange(inputs).getTypes(),
+                            SmallVector<Location>(inputs.size(), loc));
+  for (auto [domainArg, val] :
+       llvm::zip(domainBlock->getArguments(), clockDomainOp.getInputs())) {
+    val.replaceUsesWithIf(domainArg, [&](OpOperand &operand) {
+      return operand.getOwner()->getBlock() == domainBlock.get();
+    });
+  }
+  clockDomainOp->getRegion(0).push_back(domainBlock.release());
+
+  return clockDomainOp;
+}
+
+void ClockDomain::computeCrossingValues(SmallVectorImpl<Value> &inputs,
+                                        SmallVectorImpl<Value> &outputs) {
+  DenseSet<Value> inputSet, outputSet;
+  for (auto &op : *domainBlock) {
+    for (auto operand : op.getOperands()) {
+      auto *defOp = operand.getDefiningOp();
+      if (!defOp || defOp->getBlock() != domainBlock.get()) {
+        if (inputSet.insert(operand).second)
+          inputs.push_back(operand);
+      }
+    }
+    for (auto result : op.getResults()) {
+      if (llvm::any_of(result.getUsers(), [&](auto *user) {
+            return user->getBlock() != domainBlock.get();
+          })) {
+        if (outputSet.insert(result).second)
+          outputs.push_back(result);
+      }
+    }
```

Comment:
Is this sufficient for nested ops like `scf.if`, or does this possibly need an `op.walk(...)` to truly find all the value uses?

_Update:_ Just saw that you're rejecting ops with regions further down. That's a good solution :+1:

Reply:
yea I'm trying to not over-generalize here, but we might be able to pull in ops with regions when they don't have any clocked op in the region or clocked ops with all the same clock. But when there are clocked ops with different clock in a region we need to separate them somehow and that depends on the semantics of the region/op. I thought it's a good approach to keep the initial PR simple and add these cases in a later PR once the need arises.